<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
</head>
<body>
    <script>
        /***
        * 1、Promise的含义
        * promise是异步编程的一种解决方案，比传统的解决方案--回调函数和事件--更
        * 合理和更强大。他由社区最早提出和实现，ES6将其写进了语言标准，统一了用法
        *原生提供了Promise对象。
        *
        * 所谓Promise,简单说就是一个容器，里面保存着某个未来才会结束的事件（通常是
        * 一个一步操作）的结果。从语法上来说，Promise是一个对象，他可以获取异步操作
        * 的消息。Promise提供统一的API,各种异步操作都可以用同样的方法进行处理。
        *
        * Promise对象有以下两个特点
        *
        * (1) 对象的状态不受外界的影响。Promise对象代表一个异步操作，有三种状态：
        * pending(进行中)，fulfilled(已成功)，rejected(已失败)，只有异步操作的结果
        * 可以决定当前是哪一种状态，任何其他操作都无法改变这个状态，这也是Promise这个
        * 名字的由来，他的意思就是"承诺",表示其他手段无法改变。
        *
        * (2)一旦状态改变，就不会在变，任何时候都可以得到这个结果。Promise
        * 对象的状态改变，只有两种可能：从pending变为rejected和从pending变为fulfilled
        * 只要这两种情况发生，状态就凝固了，不会再变了，会一直保持这个结果，这时就称为
        * resolved(已定型)。如果改变已经发生了，你在对Promise对象添加回调函数，也会
        * 立即得到这个结果，这与事件完全不同，事件的特点是，如果你错过了，再去监听它
        * 是得不到结果的
        * 
        * 注意，为了行文方便，本章后面的resolved统一只指fulfilled状态，不包含rejected状态
        * 
        * 有了Promise对象，就可以将异步操作以同步操作的流程表达出来，避免了层层嵌套的回调函数
        * 此外，promise提供统一的接口，使得控制异步操作更加容易。
        *
        * Promise也有一些缺点，首先，无法取消Promise,一旦新建它就会立即执行，无法中途取消
        * 其次，如果不设置回调函数，Promise内部抛出的错误，不会反映到外部。第三，当处于pending
        * 状态时，无法得知目前进展到那一阶段(刚刚开始还是即将完成)
        */
    </script>

    <script>
        /**
        * 2、基本用法
        * ES6规定，Promise是一个构造函数，用来生成Promise实例
        *
        * 下面代码创造一个Promise实例
        */
        {
            const promise = new Promise((resolve,reject)=> {
                if(true) {
                    resolve();
                } else {
                    reject();
                }
            })
        }
        /**
        * Promise构造函数接受一个函数作为参数，该函数的两个参数分别是resolve和reject
        * 他们是两个函数，由JavaScript引擎提供，不用自己部署。
        * 
        * resolve 函数的作用是，将Promise对象的状态从未完成变成成功(即从pending变为rrsolved)
        * 在异步操作成功时调用，并将异步操作的结果，作为参数传递出去;reject函数的作用是
        * 将Promise对象的状态从'未完成'变为‘失败’(即从pending变为rejected),在异步操作失败
        * 时调用，并将异步操作报出的错误，作为参数传递出去。
        * 
        * Promise实例生成以后，可以使用then方法分别指定resolved状态和rejected状态的回调函数
        */
        {
            const promise = new Promise((resolve,reject) => {
                if (true) {
                    resolve();
                } else {
                    reject();
                }
            })

            promise.then(value=> {},error => {});
        }
        /**
         * then方法可以接受两个回调函数作为参数。第一个回调函数是Promise对象的状态变为
         * resolved时调用，第二个回调函数是Promise对象状态变为rejected时调用。其中，
         * 第二个函数是可选的，不一定要提供.这两个函数都接受Promise对象传出的值作为参数
         * 
         * 下面是一个Promise对象的简单例子
        */
        {
            function timeout(ms) {
                return new Promise((resolve,reject) => {
                    setTimeout(()=> {
                        return resolve()
                    },ms)
                })
            }

            timeout(100).then(value=>{

            },error=>{

            });
        }
        /**
         * 上面代码中，timeout方法返回一个Promise实例，表示表示一段时间以后才会发生
         * 的结果，过了指定的事件(ms参数)以后，Promise实例的状态变为resolved,就会触发
         * then方法绑定的回调函数
         * 
         * Promise新建后就会立即执行。
        */
        {
            let promise = new Promise((resolve,reject) => {
                console.log('Promise');
                resolve()
            });

            promise.then(value => {
                console.log('resolved');
            })

            console.log("Hi");
            // Promise
            // Hi
            // resolved
        }
        /**
         * 上面代码中，Promise新建后立即执行，所以首先输出的是Promise,然后，then
         * 方法指定的回调函数，将在当前脚本所有同步任务执行完才会执行，所以resolved
         * 最后输出。
         * 
         * 下面是异步加载图片的例子
        */
         {
             function loadImageAsync(url) {
                 return new Promise((resolve,reject) => {
                     const image = new Image();

                     image.onload = () => {
                         reslve(image)
                     };

                     image.onerror = () => {
                         reject(new Error('Could not load image at' + url));
                     }

                     image.src = url;
                 })
             }
         }
         /**
         * 上面代码中，使用Promise包装了一个图片加载的异步操作。如果加载成功，就调用
         * resolve方法，否则就调用reject
         *
         * 下面是一个用Promise对象实现的Ajax操作的例子
         */
         {
             const getJSON = function(url) {
                 const promise = new Promise((resolve,reject) => {
                     const handler = function () {
                         if (this.readyState !== 4) {
                             return;
                         }
                         if (this.status === 200) {
                             resolve(this.response);
                         } else {
                             reject(new Error(this.statusText));
                         }
                     };

                     const client = new XMLHttpRequest();
                     client.open('GET',url);
                     client.onreadystatechange = handler;
                     client.responseType = "json";
                     client.setRequestHeader('Accept',"application/json");
                     client.send();
                 })

                 return promise;
             }

             getJSON('/posts.json').then(value => {
                 console.log('Contents:' + json);
             },error => {
                 console.log('出错了',error);
             })
         }
         /**
          * 上面代码中，getJSON是对XMLHttpRequest对象的封装，用于发出一个针对
          * JSON数据的HTTP请求，并且返回一个Promise对象。需要注意的是，在getJSON
          * 内部，resolve函数和reject函数调用时，都带有参数。
          * 
          * 如果调用resolve函数和reject函数时带有参数。那么它们的参数会被传递给
          * 回调函数。reject函数的参数通常是Error对象的实例，表示抛出的错误，resolve
          * 函数的参数除了正常的值以外，还可能是另一个Promise实例，比如像下面这样。
         */
         {
             const p1 = new Promise((resolve,reject) => {

             });

             const p2 = new Promise( (resolve,reject) => {
                 resolve(p1)
             })
         }
         /**
          * 上面代码中，p1和p2都是Promise的实例，但是p2的resolve方法将p1作为参数
          * 即一个异步操作的结果是返回另一个异步操作
          * 
          * 注意，这时p1的状态就会传递给p2,也就是说，p1的状态决定了p2的状态。如果p1的
          * 状态是pending,那么p2的回调函数就会等待p1的状态改变；如果p1的状态已经是resolved
          * 或者rejected，那么p2的回调函数将会立即执行。
         */
        {
            const p1 = new Promise( (resolve,reject) => {
                setTimeout( () => reject(new Error('fail')),3000);
            })
            
            const p2 = new Promise( (resolve,reject) => {
                setTimeout( () => resolve(p1),1000)
            })

            p2.then(result => console.log(result)).catch( error => console.log(error))
        }
        /**
         * 上面代码中，p1是一个Promise，3秒之后变为rejected.p2的状态在一秒改变
         * resolve方法返回得是p1,由于p2返回的是另一个Promise,导致p2自己的状态无效了
         * 有p1决定p2的状态，所以，后面的then语句都变成针对后者p1,又过了两秒，p1变为rejected
         * 导致触发catch方法指定的回调函数
         * 
         * 注意，调用resolve或reject并不会终结Promise的参数的执行
        */
        {
            new Promise( (resolve,reject) => {
                resolve(1);
                console.log(2)
            }).then( result => console.log(result)) // 2  1
        }
        /**
         * 上面代码中，调用resolve(1)以后，后面的console.log(2)还是会执行，并且会首先
         * 打印出来，这是因为立即resolved的promise是在本轮事件循环的末尾执行，总是晚于本轮
         * 循环的同步任务
         * 
         * 一般来说，调用resolve或reject以后，Promise的使命就完成了，后继操作应该放到
         * then方法里面执行，而不应该直接写在resolve或reject的后面。所以，最好在他们前面
         * 加上return语句，这样就不会有以外了
        */
        {
            new Promise( (resolve,reject) => {
                return resolve(1)
                console.log(2)
            }).then(result => {
                console.log(result)
            })
        }
    </script>

    <script>
        /**
         * 3、Promise.prototype.then()
         * Promise实例具有then方法，也就是说，then方法是定义在原型对象Promise.prototype
         * 上的，它的作用是为Promise实例添加状态改变时的回调函数。前面说过，then方法的第一个
         * 参数是resolved状态的回调函数，第二个参数是rejected状态的回调函数。
         * 
         * then方法返回的是一个新的Promise实例(注意，不是原来那个Promise实例)。因此可以采用
         * 链式写法，即then方法后面在调用另一个then方法。
        */
        {
            //getJSON("/posts.json").then( json => json.post).then(post => post)
        }
        /**
         * 上面的代码使用then方法，依次指定了两个回调函数，第一个回调函数完成以后，会
         * 将返回结果作为参数，传入第二个回调函数。
         * 
         * 采用链式的then,可以指定一组按照次序调用的回调函数，这时，前一个回调函数，有可能
         * 返回的还是一个Promise对象(即有异步操作)，这时后一个回调函数，就会等到该Promise
         * 对象的状态发生变化，才会被调用。
        */
        {
            getJSON("/post.json").then(post => {
                return getJSON(post.commentURL)
            }).then( value => {console.log('resolved:',value)}, error => {
                console.log('rejected:',err);
            });
        }
        /**
         * 上面代码中，第一个then方法指定的回调函数，返回的是另一个Promise对象。这时
         * 第二个then方法指定的回调函数，就会等待这个新的Promise对象状态发生变化
         * 如果变为resolved,就调用第一个参数，如果状态变为rejected,就调用第二个参数
         * 
         * 如果采用箭头函数，上面的代码可以写的更简洁
        */
        {
            getJSON('post.json').then( post => getJSON(post.commoenURL))
            .then(comments => console.log('resolved',comments),
                  error => console.log('rejected',err));
        }
    </script>

    <script>
        /**
         * 4、Promise.prototype.catch()
         * Promise.prototype.catch方法是.then(null,rejection)的别名，用于指定
         * 发生错误时的回调函数
        */
        {
            getJSON
        }
    </script>
</body>
</html>